src/lib.rs

```diff
@@ -14,6 +14,182 @@
+pub mod error;
+pub mod primitives;
+#[cfg(test)]
+mod runtime;
+#[cfg(test)]
+mod tests;
+pub mod traits;
 
+use crate::error::BeefyClientError;
+use crate::primitives::{
+    BeefyNextAuthoritySet, KeccakHasher, MmrUpdateProof, SignatureWithAuthorityIndex, HASH_LENGTH,
+};
+use crate::traits::{AuthoritySet, MmrState, StorageRead, StorageWrite};
+use beefy_primitives::known_payload_ids::MMR_ROOT_ID;
+use codec::Encode;
+use sp_core::{ByteArray, H256};
+use sp_core_hashing::keccak_256;
+use sp_io::crypto;
+use sp_runtime::traits::Convert;
+
+use sp_std::prelude::*;
+
+pub struct BeefyLightClient<Store: StorageRead + StorageWrite> {
+    store: Store,
+}
+
+impl<Store: StorageRead + StorageWrite> BeefyLightClient<Store> {
+    /// Create a new instance of the light client
+    pub fn new(store: Store) -> Self {
+        Self { store }
+    }
+
+    /// Return a reference to the underlying store
+    pub fn store_ref(&self) -> &Store {
+        &self.store
+    }
+
+    /// This should verify the signed commitment signatures, and reconstruct the
+    /// authority merkle root, confirming known authorities signed the [`crate::primitives::Commitment`]
+    /// then using the mmr proofs, verify the latest mmr leaf,
+    /// using the latest mmr leaf to rotate its view of the next authorities.
+    pub fn ingest_mmr_root_with_proof(
+        &mut self,
+        mmr_update: MmrUpdateProof,
+    ) -> Result<(), BeefyClientError> {
+        let authority_set = self.store.authority_set()?;
+        let current_authority_set = &authority_set.current_authorities;
+        let next_authority_set = &authority_set.next_authorities;
+        let signatures_len = mmr_update.signed_commitment.signatures.len();
+        let validator_set_id = mmr_update.signed_commitment.commitment.validator_set_id;
+
+        // If signature threshold is not satisfied, return
+        if !validate_sigs_against_threshold(current_authority_set, signatures_len)
+            && !validate_sigs_against_threshold(next_authority_set, signatures_len)
+        {
+            return Err(BeefyClientError::IncompleteSignatureThreshold);
+        }
+
+        if current_authority_set.id != validator_set_id && next_authority_set.id != validator_set_id
+        {
+            return Err(BeefyClientError::InvalidMmrUpdate);
+        }
+
+        // Extract root hash from signed commitment and validate it
+        let mmr_root_vec = {
+            if let Some(root) = mmr_update
+                .signed_commitment
+                .commitment
+                .payload
+                .get_raw(&MMR_ROOT_ID)
+            {
+                if root.len() == HASH_LENGTH {
+                    root
+                } else {
+                    return Err(BeefyClientError::InvalidRootHash);
+                }
+            } else {
+                return Err(BeefyClientError::InvalidMmrUpdate);
+            }
+        };
+
+        let mmr_root_hash = H256::from_slice(&*mmr_root_vec);
```

Comment:
```suggestion
        // will panic if `mmr_root_vec.len() < 32` 
        let mmr_root_hash = H256::from_slice(&*mmr_root_vec);
```

Reply:
I have a check that exits the function if it's less than 32 bytes before this line

Reply:
oh ok nice,